dpd/src/api_server.rs

```diff
@@ -3061,6 +3070,200 @@ async fn counter_get(
+}
+
+/**
+ * Create a multicast group configuration.
+ *
+ * If no group ID is provided, one will be uniquely assigned.
+ */
+#[endpoint {
+    method = POST,
+    path = "/multicast/groups",
+}]
+async fn multicast_group_create(
+    rqctx: RequestContext<Arc<Switch>>,
+    group: TypedBody<mcast::MulticastGroupCreateEntry>,
+) -> Result<HttpResponseCreated<mcast::MulticastGroupResponse>, HttpError> {
+    let switch: &Switch = rqctx.context();
+    let entry = group.into_inner();
+
+    mcast::add_group(switch, entry)
+        .map(HttpResponseCreated)
+        .map_err(HttpError::from)
+}
+
+/**
+ * Delete all multicast groups (and associated routes).
+ */
+#[endpoint {
+    method = DELETE,
+    path = "/multicast/groups"
+}]
+async fn multicast_reset(
+    rqctx: RequestContext<Arc<Switch>>,
+) -> Result<HttpResponseDeleted, HttpError> {
+    let switch: &Switch = rqctx.context();
+
+    mcast::reset(switch)
+        .map(|_| HttpResponseDeleted())
+        .map_err(HttpError::from)
+}
+
+/**
+ * Delete a multicast group configuration by IP address.
+ */
+#[endpoint {
+    method = DELETE,
+    path = "/multicast/groups/{group_id}",
+}]
+async fn multicast_group_delete(
+    rqctx: RequestContext<Arc<Switch>>,
+    path: Path<MulticastGroupIdParam>,
+) -> Result<HttpResponseDeleted, HttpError> {
+    let switch: &Switch = rqctx.context();
+    let ident = path.into_inner().to_identifier()?;
+
+    mcast::del_group(switch, ident)
+        .map(|_| HttpResponseDeleted())
+        .map_err(HttpError::from)
+}
+
+/**
+ * Get the multicast group configuration for a given group ID.
+ */
+#[endpoint {
+    method = GET,
+    path = "/multicast/groups/{group_id}",
+}]
+async fn multicast_group_get(
+    rqctx: RequestContext<Arc<Switch>>,
+    path: Path<MulticastGroupIdParam>,
+) -> Result<HttpResponseOk<mcast::MulticastGroupResponse>, HttpError> {
+    let switch: &Switch = rqctx.context();
+    let ident = path.into_inner().to_identifier()?;
+
+    // Get the multicast group
+    mcast::get_group(switch, ident)
+        .map(HttpResponseOk)
+        .map_err(HttpError::from)
+}
+
+/**
+ * Update a multicast group configuration.
+*/
+#[endpoint {
+    method = PUT,
+    path = "/multicast/groups/{group_id}",
+}]
+async fn multicast_group_update(
+    rqctx: RequestContext<Arc<Switch>>,
+    path: Path<MulticastGroupIdParam>,
+    group_info: TypedBody<mcast::MulticastGroupUpdateEntry>,
+) -> Result<HttpResponseOk<mcast::MulticastGroupResponse>, HttpError> {
+    let switch: &Switch = rqctx.context();
+    let ident = path.into_inner().to_identifier()?;
+
+    mcast::modify_group(switch, ident, group_info.into_inner())
+        .map(HttpResponseOk)
+        .map_err(HttpError::from)
+}
+
+/**
+ * List all multicast groups.
+ */
+#[endpoint {
+    method = GET,
+    path = "/multicast/groups",
+}]
+async fn multicast_groups_list(
+    rqctx: RequestContext<Arc<Switch>>,
+) -> Result<HttpResponseOk<Vec<mcast::MulticastGroupResponse>>, HttpError> {
```

Comment:
Consider making this a paginated result, as the number of groups may be large.

Reply:
Yeah. Good point. I had this on my TODO, so will do (and it's easier now that I don't track routes separately anymore)